internal/validitywindow/validitywindow_test.go

```diff
@@ -298,6 +298,43 @@ func TestValidityWindowIsRepeat(t *testing.T) {
 	}
 }
 
+func TestValidityWindowBoundary(t *testing.T) {
+	r := require.New(t)
+
+	chainIndex := &testChainIndex{}
+	validityWindowDuration := int64(10)
+	validityWindow := NewTimeValidityWindow(&logging.NoLog{}, trace.Noop, chainIndex, func(int64) int64 {
+		return validityWindowDuration
+	})
+
+	// Create accepted genesis block
+	genesisBlk := newExecutionBlock(0, 0, []int64{1})
+	chainIndex.set(genesisBlk.GetID(), genesisBlk)
+	validityWindow.Accept(genesisBlk)
+
+	blk1 := newExecutionBlock(1, 0, []int64{validityWindowDuration})
+	blk2 := newExecutionBlock(2, validityWindowDuration, []int64{validityWindowDuration})
+
+	// Verify a timestamp at the validity window boundary
+	r.NoError(VerifyTimestamp(validityWindowDuration, 0, 1, validityWindowDuration))
+
+	// Including the first block should pass
+	r.NoError(validityWindow.VerifyExpiryReplayProtection(context.Background(), blk1))
+	chainIndex.set(blk1.GetID(), blk1)
+
+	// Verify a timestamp at the validity window boundary fails for both a processing
+	// and accepted parent.
+	// Processing:
+	r.ErrorIs(validityWindow.VerifyExpiryReplayProtection(context.Background(), blk2), ErrDuplicateContainer)
+
+	// Accepted:
+	validityWindow.Accept(blk1)
+	r.ErrorIs(validityWindow.VerifyExpiryReplayProtection(context.Background(), blk2), ErrDuplicateContainer)
+
+	// Verify that after passing the validity window, the timestamp is no longer valid
+	r.ErrorIs(VerifyTimestamp(validityWindowDuration, validityWindowDuration+1, 1, validityWindowDuration), ErrTimestampExpired)
```

Comment:
All of these look like separate test cases to me, can we make them separate tests? I don't think this is going to work well in a table-test which is why I suspect this test was written this way, but the cost of copy-and-paste is not that bad and I would prefer we did that over having a single monolithic test that tests multiple cases.

Reply:
I thought of this as the lifespan of a single tx being invalidated, but can break it up / see about making it a table test

Reply:
Added a separate table test for `VerifyTimestamp` and kept this test showing the lifespan of the tx (with an added comment).

The combination of the newly added test and the existing `VerifyExpiryReplayProtection` tests provides the same coverage (already covers processing/accepted at the boundary). Happy to remove this test if you prefer.

Reply:
https://github.com/ava-labs/hypersdk/pull/1895/commits/cef760574f00cc34eed55a9c9063090761d75f4b